packages/x-date-pickers/src/DateCalendar/DateCalendar.tsx

```diff
@@ -329,6 +313,25 @@ export const DateCalendar = React.forwardRef(function DateCalendar<TDate>(
     disableFuture,
   });
 
+  const handleValueChange = useEventCallback(
+    (newValue: TDate | null, selectionState?: PickerSelectionState) => {
+      setValue(newValue);
+      onChange?.(newValue, selectionState);
+
+      if (newValue != null) {
```

Comment:
We now do the check here to make sure that any value change causes a call to `onMonthChange` if needed